platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -1,73 +1,393 @@
+
+    type Error = ();
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        for px in pixels {
+            self.set_px(px.0.x as usize, px.0.y as usize, px.1)
+        }
+        Ok(())
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Helper tasks
+//
+// Friends to help us make things happen
+//////////////////////////////////////////////////////////////////////////////
+
+/// VCom - Once a second, update the vcom flag which is sent in every message.
+///
+/// TODO(AJM): The Beepberry uses a GPIO to toggle this rather than a bit in the
+/// SPI messages. This struct should be expanded to toggle the EXT pin at this
+/// rate instead, possibly as an optional field.
+struct VCom {
+    kernel: &'static Kernel,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl VCom {
+    pub async fn run(self) {
+        loop {
+            self.kernel.sleep(Duration::from_secs(1)).await;
+            let mut c = self.ctxt.lock().await;
+            c.vcom = !c.vcom;
+        }
+    }
+}
+
+/// Drawing task
+///
+/// This task draws whenever there are pending (or "dirty") changes to the frame
+/// buffer, or at a base rate of 2Hz. At the moment, we only do a full redraw to
+/// ensure that VCOM is updated periodically. In the future, we can likely send
+/// a "no-op" command instead of a full frame redraw if the dirty flag has not
+/// been set, as a power optimization or latency improvement.
+struct Draw {
+    kernel: &'static Kernel,
+    buf: FixedVec<u8>,
+    spim: SpiSenderClient,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl Draw {
+    async fn run(mut self) {
+        loop {
+            let c = self.ctxt.lock().await;
+            // render into
+            let vc = 0x01 | if c.vcom { 0x02 } else { 0x00 };
+            self.buf.as_slice_mut()[0] = 0x01 | vc;
+
+            let out_lines = self.buf.as_slice_mut()[1..].chunks_exact_mut(52);
+            let in_lines = c.sdisp.frame.iter();
+            let lines = out_lines.zip(in_lines);
+
+            for (line, (oline, iline)) in &mut lines.enumerate() {
+                oline[0] = (line as u8) + 1;
+                oline[1..51].copy_from_slice(iline);
+            }
+
+            self.buf = self.spim.send_wait(self.buf).await.map_err(drop).unwrap();
+            drop(c);
+            let _ = self
+                .kernel
+                .timeout(Duration::from_millis(500), DIRTY.wait())
+                .await;
+        }
+    }
+}
+
+/// This task is spawned by the call to [`SharpDisplay::register`]. It is a single
+/// async function that will process requests, and periodically redraw the
+/// framebuffer.
+struct CommanderTask {
+    cmd: KConsumer<Message<EmbDisplayService>>,
+    display_info: DisplayInfo,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl CommanderTask {
+    /// The entrypoint for the driver execution
+    async fn run(mut self) {
```

Comment:
nit, take it or leave it: maybe worth giving this a tracing span?

Reply:
also a good call